Tools/build/check_warnings.py

```diff
@@ -9,17 +9,42 @@
 from pathlib import Path
 
 
-def extract_warnings_from_compiler_output(compiler_output: str) -> list[dict]:
+def extract_warnings_from_compiler_output_clang(
+    compiler_output: str,
+) -> list[dict]:
     """
-    Extracts warnings from the compiler output when using
-    -fdiagnostics-format=json
+    Extracts warnings from the compiler output when using clang
+    """
+    # Regex to find warnings in the compiler output
+    clang_warning_regex = re.compile(
+        r"(?P<file>.*):(?P<line>\d+):(?P<column>\d+): warning: (?P<message>.*)"
+    )
+    compiler_warnings = []
+    for line in compiler_output.splitlines():
+        match = clang_warning_regex.match(line)
+        if match:
+            compiler_warnings.append(
+                {
+                    "file": match.group("file"),
+                    "line": match.group("line"),
+                    "column": match.group("column"),
+                    "message": match.group("message"),
+                }
+            )
+
+    return compiler_warnings
 
-    Compiler output as a whole is not a valid json document, but includes many
-    json objects and may include other output that is not json.
+
+def extract_warnings_from_compiler_output_json(
+    compiler_output: str,
+) -> list[dict]:
     """
+    Extracts warnings from the compiler output when using -fdiagnostics-format=json
 
-    # Regex to find json arrays at the top level of the file
-    # in the compiler output
+    Compiler output as a whole is not a valid json document, but includes many json
+    objects and may include other output that is not json.
+    """
+    # Regex to find json arrays at the top level of the file in the compiler output
     json_arrays = re.findall(
         r"\[(?:[^\[\]]|\[(?:[^\[\]]|\[[^\[\]]*\])*\])*\]", compiler_output
```

Comment:
PyCharm says these are redundant character escapes '\]', I've not tested it without though:

```suggestion
        r"\[(?:[^\[\]]|\[(?:[^\[\]]|\[[^\[\]]*])*])*]", compiler_output
```

Reply:
I was able to find an even simpler regex that captures the warnings in my sample outputs. Let me know if you have any other improvements.